transition.go

```diff
@@ -1,49 +1,89 @@
+var NopAction = func() {}
+
 type Transition interface {
-	NextState() State
+	execute(fromStateEngine *StateEngine) *StateEngine
 }
 
-// DirectTransition represents a straightforward transition where the next state is known a priori.
-type DirectTransition struct {
-	next State
+type ExternalTransition struct {
+	event         Event
+	toStateEngine *StateEngine
+	action        Action
 }
 
-// NewDirectTransition creates a new direct transition using the supplied state as the destination state.
-func NewDirectTransition(next State) *DirectTransition {
-	return &DirectTransition{
-		next: next,
+func NewExternalTransition(event Event, toStateEngine *StateEngine, action Action) *ExternalTransition {
+	return &ExternalTransition{
+		event:         event,
+		toStateEngine: toStateEngine,
+		action:        action,
 	}
 }
 
-// NextState returns the saved next state.
-func (dt *DirectTransition) NextState() State {
-	return dt.next
-}
+func (t *ExternalTransition) execute(fromStateEngine *StateEngine) *StateEngine {
+	// Call OnExit starting with the fromStateEngine and all parentStateEngines
+	// until parentStateEngine is nil or equal to the toState's parentStateEngine
+	parentStateEngine := fromStateEngine.GetState().OnExit(t.event)
+	for parentStateEngine != nil &&
+		(t.toStateEngine.GetParentStateEngine() == nil ||
+			parentStateEngine.Name() != t.toStateEngine.GetParentStateEngine().Name()) {
+		parentStateEngine = parentStateEngine.GetState().OnExit(t.event)
+	}
 
-// ConditionalEvaluator is a function that will return the desired state using its own evaluation criteria.
-type ConditionalEvaluator func() State
+	// Execute action on transition
+	t.action()
 
-// ConditionalTransition represents a transition where the outcome state depends on a method evaluation.
-type ConditionalTransition struct {
-	next ConditionalEvaluator
+	// Enter toState and return new currentState
+	return t.toStateEngine.GetState().OnEnter(t.event)
 }
 
-// NewConditionalTransition creates a new conditional transition using the supplied evaluator method.
-func NewConditionalTransition(next ConditionalEvaluator) *ConditionalTransition {
-	return &ConditionalTransition{
-		next: next,
+type InternalTransition struct {
+	event  Event
+	action Action
+}
+
+func NewInternalTransition(event Event, action Action) *InternalTransition {
+	return &InternalTransition{
+		event:  event,
+		action: action,
 	}
 }
 
-// NextState returns the calculated next state.
-func (ct *ConditionalTransition) NextState() State {
-	return ct.next()
+func (t *InternalTransition) execute(fromStateEngine *StateEngine) *StateEngine {
+	// Execute action on transition only (OnEnter and OnExit are not called
+	// when executing an internal transition
+	t.action()
+
+	return fromStateEngine
 }
 
-// EndTransition is a convenience variable representing a transition to an end state.
-var EndTransition = &DirectTransition{
-	next: EndState,
+type DaveEndTransition struct {
```

Comment:
what is a "Dave" and why is it in the library? ;)

Reply:
Fixed.  This was left over from an earlier refactor.